File: OpenMP/Fortran/04 Atomic Counter.f90
PROGRAM loop_decompose

  USE omp_lib
  IMPLICIT NONE
  INTEGER, PARAMETER :: max_its = 10000
  INTEGER :: i, its_global

  its_global = 0
!$OMP PARALLEL DO
  DO i = 1, max_its
!This tells the compiler that this code should be atomic
!That means that individual operations are guaranteed to complete
!in a way that can't be interrupted by other threads
!$OMP ATOMIC
    its_global = its_global + 1
!$OMP END ATOMIC
  END DO
!$OMP END PARALLEL DO

PRINT '(A,I0,A)','Counter records ', its_global, ' iterations'

END PROGRAM loop_decompose

File: OpenMP/C/04 Atomic Counter.c
#include<stdio.h>
#include<stdlib.h>
#include<omp.h>

#define MAX_ITS 10000

int main(int argc, char **argv)
{

  int i, its_global;

its_global = 0;
#pragma omp parallel
{
#pragma omp for
  for (i = 0; i< MAX_ITS; ++i){
/*This tells the compiler that this code should be atomic
That means that individual operations are guaranteed to complete
in a way that can't be interrupted by other threads*/
#pragma omp atomic update
    its_global++;
  }
}

printf( "Counter records %i iterations\n", its_global);
}
